cmd/iam-keys-check/main.go

```diff
@@ -0,0 +1,229 @@
+	s[key] = struct{}{}
+}
+
+func (s stringSet) toArray(sortValues bool) []string {
+	arr := sort.StringSlice(make([]string, 0, len(s)))
+	for k := range s {
+		arr = append(arr, k)
+	}
+	if sortValues {
+		arr.Sort()
+	}
+	return arr
+}
+
+func newstringSet() stringSet {
+	return stringSet(map[string]struct{}{})
+}
+
+// Options are the command line options
+type Options struct {
+	Profile         string `short:"p" long:"profile" description:"The AWS profile to use." required:"false" env:"AWS_PROFILE"`
+	Region          string `long:"region" description:"The AWS region to use." required:"false" env:"REGION"`
+	Lambda          bool   `short:"l" long:"lambda" description:"Run as an AWS lambda function." required:"false" env:"LAMBDA"`
+	MaxDays         uint   `long:"days" description:"The number of days." default:"90" env:"MAX_DAYS"`
+	SlackWebhookURL string `long:"slack-webhook-url" description:"The Slack Webhook Url." default:"" env:"SLACK_WEBHOOK_URL"`
+	SlackChannel    string `long:"slack-channel" description:"The Slack channel." default:"" env:"SLACK_CHANNEL"`
+}
+
+var options Options
+var logger *zap.Logger
+
+func rowToMap(header []string, row []string) map[string]string {
+	m := map[string]string{}
+	for i, h := range header {
+		m[strings.ToLower(h)] = row[i]
+	}
+	return m
+}
+
+func getCredentialReport(iamClient *iam.IAM, tries int) (*iam.GetCredentialReportOutput, error) {
+	if tries <= 0 {
+		return &iam.GetCredentialReportOutput{}, errors.New("maxmimum number of tries to get credential report reached")
+	}
+	report, err := iamClient.GetCredentialReport(&iam.GetCredentialReportInput{})
+	if err != nil {
+		if aerr, ok := err.(awserr.Error); ok {
+			if aerr.Code() == iam.ErrCodeCredentialReportNotPresentException || aerr.Code() == iam.ErrCodeCredentialReportExpiredException {
+				_, err = iamClient.GenerateCredentialReport(&iam.GenerateCredentialReportInput{})
+				if err != nil {
+					if aerr, ok = err.(awserr.Error); ok {
+						if aerr.Code() == iam.ErrCodeLimitExceededException {
+							time.Sleep(15000 * time.Millisecond)
+							return getCredentialReport(iamClient, tries-1)
+						}
+					}
+					log.Fatal(err)
+				}
+			} else if aerr.Code() == iam.ErrCodeCredentialReportNotReadyException {
+				time.Sleep(15000 * time.Millisecond)
+				return getCredentialReport(iamClient, tries-1)
+			}
+		}
+		log.Fatal(err)
+	}
+	return report, nil
+}
+
+func triggerCheck() {
+
+	maxDays := float64(options.MaxDays)
+
+	if maxDays <= 0 {
+		log.Fatal("MaxDays must be greater than 0.")
+	}
+
+	if len(options.SlackWebhookURL) == 0 {
+		log.Fatal("SlackWebhookURL is blank.")
+	}
+
+	if len(options.SlackChannel) == 0 {
+		log.Fatal("SlackChannel is blank.")
+	}
+
+	sess := session.MustMakeSession(options.Region, options.Profile)
+	iamClient := iam.New(sess)
+
+	report, err := getCredentialReport(iamClient, 5)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	reader := csv.NewReader(strings.NewReader(string(report.Content)))
```

Comment:
I think you can use https://docs.aws.amazon.com/sdk-for-go/api/service/iam/#GetCredentialReportOutput.GoString for turning the report in to a string.

Reply:
GoString doesn't return what you think it does.  It returns:

```
{
  Content: <binary> len 5289,
  GeneratedTime: 2018-07-23 18:25:05 +0000 UTC,
  ReportFormat: "text/csv"
}
```